### PR TITLE
Remove centralised port allocation for tests

### PR DESCRIPTION
**Pull request by tcsc**
Ports used by the unit tests have been allocated by pulling them out of a list, with no guarantee that the port is not actually in use. This central allocation point means that tests cannot be split into separate packages to be run in parallel, as the ports allocated between the various packages will be allocated multiple times  and end up intermittently clashing.
    
There is also no guarantee, even when the tests are run serially, that the ports will not clash with services already running on the machine.
    
This patch (largely) replaces the use of this centralised port allocation with pre-created listeners injected into the test via the file descriptor import mechanism use by Teleport to pass open ports to child processes.

There are still some cases where the old port allocation system is still in use. I felt this was already getting beyond the bounds of sensibly reviewable, so I have left those for a further PR after this.

**Note for reviewers:**

I would review the changes to the `integrations/helpers` package first, as that will inform a lot of the other changes. Start there with `ports.go` and work your way back.

See-Also: #12421
See-Also: #14408

### CODE REVIEW

**Comment by tcsc** [comment]
Known failing tests: 

TestALPNProxyHTTPProxyNoProxyDial
TestALPNSNIHTTPSProxy
TestALPNSNIProxyDatabaseAccess
TestALPNSNIProxyDatabaseAccess/mongo
TestALPNSNIProxyDatabaseAccess/mongo/connect_to_leaf_cluster_via_proxy
TestALPNSNIProxyDatabaseAccess/mongo/connect_to_main_cluster_via_proxy
TestAppServersHA/LeafServer
TestAppServersHA/RootServer
TestIntegrations/HA
TestIntegrations/ListResourcesAcrossClusters
TestIntegrations/PAM/Enabled_with_custom_environment
TestIntegrations/PAM/Enabled_with_Module_&_Session_functions_succeeding
TestIntegrations/PAM/Enabled_with_Module_Account_&_Session_functions_succeeding
TestIntegrations/Shutdown
TestIntegrations/TwoClustersProxy
TestIntegrations/TwoClustersTunnel/node
TestIntegrations/TwoClustersTunnel/proxy
TestMultiPortHTTPSProxy

**Comment by greedy52** [comment]
Just for tracking purpose, I've seen these service tests fail for the same problem
https://github.com/gravitational/teleport/blob/3ffe935f478e51a339257464e5de0f0e28e398b6/lib/service/service_test.go#L702-L714

File: integration/helpers/ports.go [hunk 1]
```diff
@@ -38,168 +41,207 @@ func init() {
 	}
 }
 
+// NewPortValue is deprecated. Do not use in new code. Use NewListener() and
+// friends instead.
```
Comment (strideynet) [inline integration/helpers/ports.go:45]: Total nit-pick, [but if you use `Deprecated: ` then it'll show as deprecated in various editors and godoc.](https://rakyll.org/deprecated/)

File: integration/integration_test.go [hunk 2]
```diff
@@ -1140,15 +1126,33 @@ func testEscapeSequenceYesTrigger(t *testing.T, terminal *Terminal, sess <-chan
...
-	terminal.Type("\aecho hi\n\r")
+	terminal.Type("\aecho made it to here!\n\r")
 
+	// Expect that the session will NOT be disconnected by the escape sequence,
+	// and so the arbitrary text will eventually end up in the terminal buffer.
 	require.Eventually(t, func() bool {
-		// if the session didn't end, we should see the output of the last write
-		return strings.Contains(terminal.Output(1000), "hi")
+		select {
+		case err := <-sess:
```
Comment (strideynet) [inline integration/integration_test.go:1142]: Nice fix here.

**Review by strideynet — APPROVED** [review]
Extremely glad to finally see this up for review, a key component in hopefully getting our tests running a little faster (and hopefully failing less :) )

Do we expect any major sticking points in converting the remaining port allocations to use this new technique ?

File: integration/proxy_helpers_test.go [hunk 3]
```diff
@@ -48,12 +43,18 @@ import (
...
+	"github.com/gravitational/trace"
 	"github.com/jackc/pgconn"
 	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
+// 	v1 "k8s.io/api/core/v1"
```
Comment (tcsc) [inline integration/proxy_helpers_test.go:55]: Delete

File: integration/helpers/instance.go [hunk 4]
```diff
@@ -345,7 +346,7 @@ func NewInstance(cfg InstanceConfig) *TeleInstance {
 		Cert:       cert,
 		TLSCACert:  tlsCACert,
 		TLSCert:    tlsCert,
-		TunnelAddr: net.JoinHostPort(cfg.NodeName, i.GetPortReverseTunnel()),
+		TunnelAddr: i.ReverseTunnel, // TODO(tcsc): net.JoinHostPort(cfg.NodeName, i.GetPortReverseTunnel()),
```
Comment (r0mant) [inline integration/helpers/instance.go:349]: Is this TODO item still relevant? I would expand it if so, right now it's not clear what the actual TODO here is.

File: integration/helpers/ports.go [hunk 5]
```diff
@@ -38,168 +41,207 @@ func init() {
...
 	}
-	return net.JoinHostPort(i.Host, i.GetPortMySQL())
 }
 
-func (i *InstancePorts) GetReverseTunnelAddr() string {
-	if i.ReverseTunnel == nil {
-		return ""
-	}
-	return net.JoinHostPort(i.Host, i.GetPortReverseTunnel())
+func PortStr(t *testing.T, addr string) string {
```
Comment (r0mant) [inline integration/helpers/ports.go:177]: Nit: Add godocs on public methods.

File: integration/helpers/ports.go [hunk 6]
```diff
@@ -38,168 +41,207 @@ func init() {
...
+// parent and child processes) to inject preconfigured listeners to Teleport
+// instances under test. The ports are allocated and bound at runtime, so there
+// should be no issues with port clashes on parallel tests.
+//
+// The resulting file descriptor is added to the `fds` slice, which can then be
+// given to a teleport instance on startup in order to suppl
+func NewListenerOn(t *testing.T, hostAddr string, ty service.ListenerType, fds *[]service.FileDescriptor) string {
+	t.Helper()
+
+	l, err := net.Listen("tcp", hostAddr+":0")
```
Comment (r0mant) [inline integration/helpers/ports.go:209]: Nit: Use net.JoinHostPort?

File: integration/integration_test.go [hunk 7]
```diff
@@ -1950,9 +1973,18 @@ func testHA(t *testing.T, suite *integrationTestSuite) {
...
+	// fail when it attempts to start back up again. We can't just inject a
+	// totally new listener config into SiteA when it restarts, or SiteB won't
+	// be able to  find it.
+	//
+	// The least bad option is to duplicate all of SiteA's Listener FDs and
+	// inject those duplicates prior to restarting the SiteA cluster.
+	aFdCache, err := a.Process.ExportFileDescriptors()
+	require.NoError(t, err)
+
+	sshPort, _, _ := a.StartNodeAndProxy(t, "cluster-a-node")
```
Comment (r0mant) [inline integration/integration_test.go:1987]: Is the error ignored here?

File: integration/integration_test.go [hunk 8]
```diff
@@ -2094,9 +2127,7 @@ func testMapRoles(t *testing.T, suite *integrationTestSuite) {
 	tryCreateTrustedCluster(t, aux.Process.GetAuthServer(), trustedCluster)
 	waitForTunnelConnections(t, main.Process.GetAuthServer(), clusterAux, 1)
 
-	nodePorts := helpers.NewPortSlice(3)
-	sshPort, proxyWebPort, proxySSHPort := nodePorts[0], nodePorts[1], nodePorts[2]
-	require.NoError(t, aux.StartNodeAndProxy("aux-node", sshPort, proxyWebPort, proxySSHPort))
+	sshPort, _, _ := aux.StartNodeAndProxy(t, "aux-node")
```
Comment (r0mant) [inline integration/integration_test.go:2130]: Is the error ignored, or are the 2nd and 3rd arguments something else?

**Review by r0mant — APPROVED** [review]
Pretty hard to evaluate all the changes here, but I'd say if this improves the situation and there are no failing tests, let's get this in.

File: integration/integration_test.go (hunk 7 above)
Comment (tcsc) [reply integration/integration_test.go:1987]: Amy errors are caught inside `StartNodeAndProxy()` and result in an immediately failed test.

File: integration/integration_test.go (hunk 8 above)
Comment (tcsc) [reply integration/integration_test.go:2130]: Something else. Any errors are caught inside `StartNodeAndProxy()` and result in an immediately failed test.

File: integration/helpers/ports.go (hunk 1 above)
Comment (tcsc) [reply integration/helpers/ports.go:45]: Done

File: integration/helpers/instance.go (hunk 4 above)
Comment (tcsc) [reply integration/helpers/instance.go:349]: Not relevant any more. Gone.

File: integration/proxy_helpers_test.go (hunk 3 above)
Comment (tcsc) [reply integration/proxy_helpers_test.go:55]: Done

File: integration/helpers/ports.go (hunk 6 above)
Comment (tcsc) [reply integration/helpers/ports.go:209]: Done

**Pull request merged**